### PR TITLE
useMutation: fix rules of React violations

### PR DESCRIPTION
**Pull request by phryneas**
Another one for #11511...

### CODE REVIEW

File: src/react/hooks/useMutation.ts [hunk 1]
```diff
@@ -221,17 +219,22 @@ export function useMutation<
 
   const reset = React.useCallback(() => {
     if (ref.current.isMounted) {
-      const result = { called: false, loading: false, client };
+      const result = {
+        called: false,
+        loading: false,
+        client: ref.current.client,
```
Comment (phryneas) [inline src/react/hooks/useMutation.ts:225]: Woops 🙃

**Review by jerelmiller — APPROVED** [review]
Looks good to me!

**Comment by mikehdt** [comment]
@phryneas This PR has introduced a `React.useLayoutEffect` call, which causes a slew of warnings when it gets rendered server-side.

You might need to consider guarding against this with some of the solutions suggested at https://reactjs.org/link/uselayouteffect-ssr

**Comment by Multiply** [comment]
> @phryneas This PR has introduced a `React.useLayoutEffect` call, which causes a slew of warnings when it gets rendered server-side.
> 
> You might need to consider guarding against this with some of the solutions suggested at https://reactjs.org/link/uselayouteffect-ssr

We're seeing a lot of warnings after the introduction of useLayoutEffect, so we downgraded to 3.10.4 again.

**Comment by phryneas** [comment]
@mikehdt @Multiply @quadrifolia thank y'all for the report! I'll see that I get a patch for this out at some point today or tomorrow!

**Comment by phryneas** [comment]
@mikehdt @Multiply @quadrifolia could you please try out this build and report back if that solves the problem?

```sh
npm i @apollo/client@0.0.0-pr-11900-20240620115719
```

**Comment by Multiply** [comment]
> @mikehdt @Multiply @quadrifolia could you please try out this build and report back if that solves the problem?
> 
> ```shell
> npm i @apollo/client@0.0.0-pr-11900-20240620115719
> ```

The error seems to be gone with that version of @apollo/client indeed.

**Comment by phryneas** [comment]
@mikehdt @Multiply @quadrifolia I just released 3.10.6 which should address your problem.

**Pull request merged**